File: 과제3_stube/stube.cpp
/*#include <bits/stdc++.h>

using namespace std;

class Coordinate {
public:
	long double x = 0, y = 0, z = 0;
};

long double distance(Coordinate A, Coordinate B, Coordinate P, long double t) {
	if (t >= 1) {
		return sqrt(pow(B.x - P.x, 2) + pow(B.y - P.y, 2) + pow(B.z - P.z, 2));
	}
	else if (0 < t && t < 1) {
		return (long double)sqrt((-1 * pow(((A.x - P.x) * (B.x - A.x) + (A.y - P.y) * (B.y - A.y) + (A.z - P.z) * (B.z - A.z)), 2)
			/ ((pow(B.x - A.x, 2) + pow(B.y - A.y, 2) + pow(B.z - A.z, 2)))) + pow(A.x - P.x, 2) + pow(A.y - P.y, 2) + pow(A.z - P.z, 2));
	}
	else {
		return sqrt(pow(A.x - P.x, 2) + pow(A.y - P.y, 2) + pow(A.z - P.z, 2));
	}
}

long double solve(Coordinate A, Coordinate B, Coordinate P) {
	long double location;
	location = (long double)(-1 * ((A.x - P.x) * (B.x - A.x) + (A.y - P.y) * (B.y - A.y) + (A.z - P.z) * (B.z - A.z)) / ((pow(B.x - A.x, 2) + pow(B.y - A.y, 2) + pow(B.z - A.z, 2))));
	return distance(A, B, P, location);
}

long double pickSP(Coordinate A, Coordinate B, Coordinate C, Coordinate D) {
	long double i = 0, l = 0, least = 10000000000;
	Coordinate P, Pt, T;
	vector<long double> vec;
	while (i < 1.01) {
		Pt.x = i * A.x + (1 - i) * B.x;
		Pt.y = i * A.y + (1 - i) * B.y;
		Pt.z = i * A.z + (1 - i) * B.z;
		long double d = solve(C, D, Pt);
		vec.push_back(d);
		if (least > d) {
			least = d;
			P.x = Pt.x;
			P.y = Pt.y;
			P.z = Pt.z;
			l = i;
		}
		i += 0.25;
	}
	long double min = *min_element(vec.begin(), vec.end());
	long double max = *max_element(vec.begin(), vec.end());
	if (min == 0) {
		return 0;
	}
	if (round(min * 100000000) / 100000000 == round(max * 100000000) / 100000000) {
		return least;
	}
	if (l == 1) {
		T.x = 0.75 * A.x + 0.25 * B.x;
		T.y = 0.75 * A.y + 0.25 * B.y;
		T.z = 0.75 * A.z + 0.25 * B.z;
		return pickSP(T, P, C, D);
	}
	else if (1 > l&& l >= 0.75) {
		T.x = 0.5 * A.x + 0.5 * B.x;
		T.y = 0.5 * A.y + 0.5 * B.y;
		T.z = 0.5 * A.z + 0.5 * B.z;
		return pickSP(T, A, C, D);
	}
	else if (0.25 < l && l < 0.75) {
		T.x = 0.25 * A.x + 0.75 * B.x;
		T.y = 0.25 * A.y + 0.75 * B.y;
		T.z = 0.25 * A.z + 0.75 * B.z;
		Pt.x = 0.75 * A.x + 0.25 * B.x;
		Pt.y = 0.75 * A.y + 0.25 * B.y;
		Pt.z = 0.75 * A.z + 0.25 * B.z;
		return pickSP(T, Pt, C, D);
	}
	else if (0 < l && l <= 0.25) {
		T.x = 0.5 * A.x + 0.5 * B.x;
		T.y = 0.5 * A.y + 0.5 * B.y;
		T.z = 0.5 * A.z + 0.5 * B.z;
		return pickSP(B, T, C, D);
	}
	else if (l == 0) {
		T.x = 0.25 * A.x + 0.75 * B.x;
		T.y = 0.25 * A.y + 0.75 * B.y;
		T.z = 0.25 * A.z + 0.75 * B.z;
		return pickSP(P, T, C, D);
	}
}

int main() {
	fstream inp("stube.inp");
	ofstream out("stube.out");
	Coordinate A, B, C, D;
	inp >> A.x >> A.y >> A.z >> B.x >> B.y >> B.z >> C.x >> C.y >> C.z >> D.x >> D.y >> D.z;
	inp.close();
	long double ab_to_cd, cd_to_ab, result;
	ab_to_cd = pickSP(A, B, C, D);
	cd_to_ab = pickSP(C, D, A, B);
	out << ceil(min(ab_to_cd, cd_to_ab)) << endl;
	out.close();
}*/

#include <iostream>

int main() {
	for (int i : {2, 3, 5, 7, 9}) { std::cout << i << std::endl; }
}

File: 과제3_stube/stube.py
#과제 2번 예시코드를 사용하였습니다.

import math as m
import random
import sys
sys.setrecursionlimit(10 ** 8)

class Coordinate:
    x = y = z = 0

    def __init__(self, inp):
        self.x, self.y, self.z = map(int, inp.readline().split())


def dist(x1, y1, z1, x2, y2, z2):
    return ((x1 - x2) ** 2 + (y1 - y2) ** 2 + (z1 - z2) ** 2) ** 0.5


def mid(x1, y1, z1, x2, y2, z2, x3, y3, z3):
    sx, sy, sz = (x1 + x2) / 2, (y1 + y2) / 2, (z1 + z2) / 2
    dap = dist(x1, y1, z1, x3, y3, z3)
    dbp = dist(x2, y2, z2, x3, y3, z3)
    if round(dap, 8) == round(dbp, 8):
        return dist(sx, sy, sz, x3, y3, z3)
    elif dap > dbp:
        return mid(sx, sy, sz, x2, y2, z2, x3, y3, z3)
    elif dap < dbp:
        return mid(x1, y1, z1, sx, sy, sz, x3, y3, z3)


def pickSP(x1, y1, z1, x2, y2, z2, x3, y3, z3, x4, y4, z4):
    i = l = px = py = pz = 0
    arr = []
    least = 10000000
    while i < 1.01:
        tpx, tpy, tpz = i * x1 + (1 - i) * x2, i * y1 + (1 - i) * y2, i * z1 + (1 - i) * z2
        d = mid(x3, y3, z3, x4, y4, z4, tpx, tpy, tpz)
        arr.append(d)
        if least > d:
            least = d
            px, py, pz = tpx, tpy, tpz
            l = i
        i += 0.25
    if min(arr) == 0:
        return 0
    if round(min(arr), 3) == round(max(arr), 3):
        return least
    if l == 1:
        sx, sy, sz = 0.75 * x1 + 0.25 * x2, 0.75 * y1 + 0.25 * y2, 0.75 * z1 + 0.25 * z2
        return pickSP(sx, sy, sz, px, py, pz, x3, y3, z3, x4, y4, z4)
    elif l == 0.75:
        sx, sy, sz = 0.5 * x1 + 0.5 * x2, 0.5 * y1 + 0.5 * y2, 0.5 * z1 + 0.5 * z2
        return pickSP(sx, sy, sz, x1, y1, z1, x3, y3, z3, x4, y4, z4)
    elif l == 0.5:
        sx1, sy1, sz1 = 0.25 * x1 + 0.75 * x2, 0.25 * y1 + 0.75 * y2, 0.25 * z1 + 0.75 * z2
        sx2, sy2, sz2 = 0.75 * x1 + 0.25 * x2, 0.75 * y1 + 0.25 * y2, 0.75 * z1 + 0.25 * z2
        return pickSP(sx1, sy1, sz1, sx2, sy2, sz2, x3, y3, z3, x4, y4, z4)
    elif l == 0.25:
        sx, sy, sz = 0.5 * x1 + 0.5 * x2, 0.5 * y1 + 0.5 * y2, 0.5 * z1 + 0.5 * z2
        return pickSP(x2, y2, z2, sx, sy, sz, x3, y3, z3, x4, y4, z4)
    elif l == 0:
        sx, sy, sz = 0.25 * x1 + 0.75 * x2, 0.25 * y1 + 0.75 * y2, 0.25 * z1 + 0.75 * z2
        return pickSP(px, py, pz, sx, sy, sz, x3, y3, z3, x4, y4, z4)


def Input():
    return open("stube/sampleData/9.inp"), open("stube.out", "w")


#if __name__ == "__main__":
while 1:
    inp1 = open("stube/sampleData/9.inp", "w")
    a = random.sample(range(-10, 10), 3)
    b = random.sample(range(-10, 10), 3)
    c = random.sample(range(-10, 10), 3)
    d = random.sample(range(-10, 10), 3)
    for i in a:
        inp1.write("%d " % i)
    inp1.write("\n")
    for i in b:
        inp1.write("%d " % i)
    inp1.write("\n")
    for i in c:
        inp1.write("%d " % i)
    inp1.write("\n")
    for i in d:
        inp1.write("%d " % i)
    inp1.write("\n")
    inp1.close()
    inp, out = Input()
    A = Coordinate(inp)
    B = Coordinate(inp)
    C = Coordinate(inp)
    D = Coordinate(inp)
    inp.close()
    ab_to_cd = pickSP(C.x, C.y, C.z, D.x, D.y, D.z, A.x, A.y, A.z, B.x, B.y, B.z)
    cd_to_ab = pickSP(A.x, A.y, A.z, B.x, B.y, B.z, C.x, C.y, C.z, D.x, D.y, D.z)
    if ab_to_cd is None or cd_to_ab is None:
        out.write("0")
    print(min(ab_to_cd, cd_to_ab))
    print(m.ceil(min(ab_to_cd, cd_to_ab)))
    if min(ab_to_cd, cd_to_ab) < 0.001:
        print("0")
        break
    else:
        print("%s" % m.ceil(min(ab_to_cd, cd_to_ab)))


'''def distance(x1, y1, z1, x2, y2, z2, x3, y3, z3, l):
    if l >= 1:
        return ((x2 - x3) ** 2 + (y2 - y3) ** 2 + (z2 - z3) ** 2) ** 0.5
    elif 0 < l < 1:
        return ((-1 * (((x1 - x3) * (x2 - x1) + (y1 - y3) * (y2 - y1) + (z1 - z3) * (z2 - z1)) ** 2)
                     / ((x2 - x1) ** 2 + (y2 - y1) ** 2 + (z2 - z1) ** 2))
                    + (x1 - x3) ** 2 + (y1 - y3) ** 2 + (z1 - z3) ** 2) ** 0.5
    elif l <= 0:
        return ((x1 - x3) ** 2 + (y1 - y3) ** 2 + (z1 - z3) ** 2) ** 0.5


def location(x1, y1, z1, x2, y2, z2, x3, y3, z3):
    l = (-1 * (((x1 - x3) * (x2 - x1)) + ((y1 - y3) * (y2 - y1)) + ((z1 - z3) * (z2 - z1)))
             / (((x2 - x1) ** 2) + ((y2 - y1) ** 2) + ((z2 - z1) ** 2)))
    return distance(x1, y1, z1, x2, y2, z2, x3, y3, z3, l)


def dist(x1, y1, z1, x2, y2, z2, x3, y3, z3, x4, y4, z4):
    x5, y5, z5 = 0.5 * x3 + 0.5 * x4, 0.5 * y3 + 0.5 * y4, 0.5 * z3 + 0.5 * z4
    x6, y6, z6 = 0.25 * x3 + 0.75 * x4, 0.25 * y3 + 0.75 * y4, 0.25 * z3 + 0.75 * z4
    x7, y7, z7 = 0.75 * x3 + 0.25 * x4, 0.75 * y3 + 0.25 * y4, 0.75 * z3 + 0.25 * z4
    abc = location(x1, y1, z1, x2, y2, z2, x3, y3, z3)
    abp1 = location(x1, y1, z1, x2, y2, z2, x6, y6, z6)
    abp2 = location(x1, y1, z1, x2, y2, z2, x5, y5, z5)
    abp3 = location(x1, y1, z1, x2, y2, z2, x7, y7, z7)
    abd = location(x1, y1, z1, x2, y2, z2, x4, y4, z4)
    if abc == 0 or abd == 0 or abp1 == 0 or abp2 == 0 or abp3 == 0:
        return 0
    if round(abc, 2) == round(abp1, 2) and round(abc, 2) == round(abd, 2):
        return abp1
    if abc < abp1 and abc < abp2 and abc < abp3 and abc < abd:
        return dist(x1, y1, z1, x2, y2, z2, x7, y7, z7, x3, y3, z3)
    elif abp1 < abc and abp1 < abp2 and abp1 < abp3 and abp1 < abd:
        return dist(x1, y1, z1, x2, y2, z2, x4, y4, z4, x5, y5, z5)
    elif abp2 < abc and abp2 < abp1 and abp2 < abp3 and abp2 < abd:
        return dist(x1, y1, z1, x2, y2, z2, x7, y7, z7, x6, y6, z6)
    elif abp3 < abc and abp3 < abp1 and abp3 < abp2 and abp3 < abd:
        return dist(x1, y1, z1, x2, y2, z2, x5, y5, z5, x3, y3, z3)
    elif abd < abc and abd < abp1 and abd < abp2 and abd < abp3:
        return dist(x1, y1, z1, x2, y2, z2, x4, y4, z4, x6, y6, z6)'''